### PR TITLE
updating scripts/genproto.sh

### LINKED ISSUE #13050: scripts/genproto.sh: updates needed
The API-reference generator needs a few updates to:

- [ ] Match the changes mentioned in https://github.com/etcd-io/website/issues/330.
- [ ] Generate page front matter
- [ ] Reconsider where the output should go: currently, the script is targeting `Documentation/dev-guide/api_concurrency_reference_v3.md`. The website docs aren't under `Documentation` anymore. Either the output target should be changed, or instructions should be given to let the script user know where to copy/submit the files to.
- As a followup action, port to the release-3.5 branch

cc @nate-double-u @gyuho

### PR DESCRIPTION
**Pull request by nate-double-u**
Changing API-reference generator
- [x] matching changes mentioned in https://github.com/etcd-io/website/issues/330
- [x] generating page front-matter
- [ ] reconsider where the output should go

fixes #13050
contributes to: [#111](https://github.com/etcd-io/website/issues/111), #12768

### Files generated

Most recent commit:
- [api_concurrency_reference_v3.md](https://github.com/etcd-io/etcd/files/6580135/api_concurrency_reference_v3.md)
- [api_reference_v3.md](https://github.com/etcd-io/etcd/files/6580136/api_reference_v3.md)

Previous commits:

- [api_concurrency_reference_v3.md](https://github.com/etcd-io/etcd/files/6579849/api_concurrency_reference_v3.md)
- [api_reference_v3.md](https://github.com/etcd-io/etcd/files/6579850/api_reference_v3.md)
- [api_reference_v3.md](https://github.com/etcd-io/etcd/files/6572053/api_reference_v3.md)
- [api_concurrency_reference_v3.md](https://github.com/etcd-io/etcd/files/6572054/api_concurrency_reference_v3.md)

/cc @chalin

### CODE REVIEW

File: scripts/genproto.sh [hunk 1]
```diff
@@ -95,18 +95,34 @@ run_go_tool github.com/hexfusion/schwag -input=Documentation/dev-guide/apispec/s
 if [ "$1" != "--skip-protodoc" ]; then
   log_callout "protodoc is auto-generating grpc API reference documentation..."
 
+  # API reference
   run rm -rf Documentation/dev-guide/api_reference_v3.md
```
Comment (nate-double-u) [inline scripts/genproto.sh:99]: Once we know for sure if/where we're going to move these files, I'd like to set an env var for them as we use them a couple times.

File: scripts/genproto.sh [hunk 2]
```diff
@@ -95,18 +95,34 @@ run_go_tool github.com/hexfusion/schwag -input=Documentation/dev-guide/apispec/s
...
+---
 
+This API reference is autogenerated from the named \`.proto\` files.
+''' || exit 2
+
+  # remove the first 3 lines of the doc as an empty --title adds '### ' to the top of the file.
+  run sed -i -e 1,3d Documentation/dev-guide/api_reference_v3.md
+
+  # API reference: concurrency
   run rm -rf Documentation/dev-guide/api_concurrency_reference_v3.md
```
Comment (nate-double-u) [inline scripts/genproto.sh:114]: Same as above, once we know if (and where) these file are moving, then I'd like to set an env var to help simplify this section.

**Comment by nate-double-u** [comment]
I don't think the test failures are related to the documentation updates.

File: scripts/genproto.sh [hunk 3]
```diff
@@ -95,18 +95,34 @@ run_go_tool github.com/hexfusion/schwag -input=Documentation/dev-guide/apispec/s
...
   log_callout "protodoc is auto-generating grpc API reference documentation..."
 
+  # API reference
   run rm -rf Documentation/dev-guide/api_reference_v3.md
   run_go_tool go.etcd.io/protodoc --directories="api/etcdserverpb=service_message,api/mvccpb=service_message,server/lease/leasepb=service_message,api/authpb=service_message" \
-    --title="etcd API Reference" \
     --output="Documentation/dev-guide/api_reference_v3.md" \
     --message-only-from-this-file="api/etcdserverpb/rpc.proto" \
-    --disclaimer="This is a generated documentation. Please read the proto files for more." || exit 2
+    --disclaimer='''---
```
Comment (ptabor) [inline scripts/genproto.sh:103]: I don't think there are triple quotes in bash support. 
One of workarounds is to use temporary variable: 

```
DISCLAIMER=$(cat <<-END
    This is line one.
    This is line two.
    This is line three.
END
)
```

**Comment by ptabor** [comment]
I think that the tests fail due to shellcheck not understaning tripple quotes: 
```
FAIL: (code:1):
scripts/genproto.sh:103:20: warning: Did you forget to close this single quoted string? [SC1078]
  % 'shellcheck' '-fgcc' 'build' 'test' 'scripts/build-release.sh' 'scripts/codecov_upload.sh' 'scripts/fix.sh' 'scripts/genproto.sh' 'scripts/install-marker.sh' 'scripts/release_mod.sh' 'scripts/test_lib.sh' 'scripts/update_dep.sh' 'scripts/updatebom.sh' './build.sh' './test.sh'
FAIL: 'run shellcheck -fgcc build test scripts/build-release.sh scripts/codecov_upload.sh scripts/fix.sh scripts/genproto.sh scripts/install-marker.sh scripts/release_mod.sh scripts/test_lib.sh scripts/update_dep.sh scripts/updatebom.sh ./build.sh ./test.sh' checking failed (!=0 return code)
scripts/genproto.sh:103:20: note: Expressions don't expand in single quotes, use double quotes for that. [SC2016]
scripts/genproto.sh:108:1: note: This is actually an end quote, but due to next char it looks suspect. [SC1079]
FAIL: 'shellcheck' failed at Mon May 31 20:05:35 UTC 2021

scripts/genproto.sh:117:20: warning: Did you forget to close this single quoted string? [SC1078]
scripts/genproto.sh:117:20: note: Expressions don't expand in single quotes, use double quotes for that. [SC2016]
scripts/genproto.sh:122:1: note: This is actually an end quote, but due to next char it looks suspect. [SC1079]
Error: Process completed with exit code 255.
```

**Review by chalin — COMMENTED** [review]
See inline comments, and feedback by @ptabor.

@nate-double-u - have you been able to run this script? I'm getting `sed` **errors** under macOS. Maybe this currently only runs under Linux?

**Comment by nate-double-u** [comment]
> have you been able to run this script?

The files I linked in the description were generated. I admittedly did not run it on multiple platforms though.

File: scripts/genproto.sh [hunk 4]
```diff
@@ -95,18 +95,34 @@ run_go_tool github.com/hexfusion/schwag -input=Documentation/dev-guide/apispec/s
...
+title: "API reference"
```
[the remark this answers is not in the document]
Comment (nate-double-u) [reply scripts/genproto.sh:104]: Good catch -- that's a copy-paste error

**Comment by nate-double-u** [comment]
> I'm getting sed errors under macOS. Maybe this currently only runs under Linux?

Doing a quick look around, it looks like there may be differences between `sed` on macOS and on Linux. I haven't tested the unmodified script on macOS, I'll have to test it to see if it has problems as well (or see if it's the additional line I added that's causing the issue)

File: scripts/genproto.sh (hunk 3 above)
Comment (nate-double-u) [reply scripts/genproto.sh:103]: Thanks for the workaround, I'll add that in.

Comment (nate-double-u) [reply scripts/genproto.sh:103]: Interesting, the `'''` is working with my bash (GNU bash, version 5.0.17(1)-release), but the `cat` method you shared seems to split the string into individual arguments when passed into `--disclaimer`.

Are you able to run the script as committed @ptabor?

I'm going to look closer at `shellcheck` as well (I've used it before on other projects, but don't have it set up for this project yet).

**Comment by nate-double-u** [comment]
> I think that the tests fail due to shellcheck not understaning tripple quotes

Thanks @ptabor, I'll take a closer look at that

**Comment by nate-double-u** [comment]
Files generated by the most recent commit:
- [api_concurrency_reference_v3.md](https://github.com/etcd-io/etcd/files/6579849/api_concurrency_reference_v3.md)
- [api_reference_v3.md](https://github.com/etcd-io/etcd/files/6579850/api_reference_v3.md)

**Comment by nate-double-u** [comment]
I think this last commit clears the `shellcheck` error. The remaining failed test seems unrelated.

**Review by chalin — COMMENTED** [review]
Hmm, you probably shouldn't be modifying `go.mod` and `go.sum`. (Yeah, I know that's what happens when we get the tools we need to run this script.)

I'm thinking that is makes more sense to have this **script reside in the etcd website repo**. 

Thoughts @ptabor @spzala @gyuho?

**Comment by nate-double-u** [comment]
> Hmm, you probably shouldn't be modifying `go.mod` and `go.sum`. (Yeah, I know that's what happens when we get the tools we need to run this script.)

Ya i just noticed that and have removed them.

**Comment by nate-double-u** [comment]
Rebased to clean up commits.

Latest force push adds env vars, new console output on success:
```
protodoc is finished.

To publish the API References, copy the following files
  - Documentation/dev-guide/api_reference_v3.md
  - Documentation/dev-guide/api_concurrency_reference_v3.md
to the etcd-io/website repo under the /content/en/docs/next/dev-guide/ folder.
(https://github.com/etcd-io/website/tree/main/content/en/docs/next/dev-guide)

./genproto SUCCESS
```

**Comment by nate-double-u** [comment]
> I'm thinking that it makes more sense to have this **script reside in the etcd website repo**.

If we don't move it, do we want to consider building a GitHub action or something to build the references and push the files across to the etcd-io/website repo automatically? (I've found systems that automate this sort of thing to be fairly fragile in past, but does have the advantage of not requiring someone to remember to update the files manually).

File: scripts/genproto.sh [hunk 5]
```diff
@@ -92,23 +92,46 @@ done
...
+title: \"API reference: concurrency\"
+---
+
+This API reference is autogenerated from the named \`.proto\` files." || exit 2
+
+  # remove the first 3 lines of the doc as an empty --title adds '### ' to the top of the file.
+  run sed -i -e 1,3d ${API_REFERENCE_CONCURRENCY_FILE}
 
   log_success "protodoc is finished."
+  log_success ""
```
Comment (ptabor) [inline scripts/genproto.sh:129]: Maybe:

log_warning "The files has NOT been automatically published on the website."

**Review by ptabor — APPROVED** [review]

**Review by chalin — APPROVED** [review]

**Comment by chalin** [comment]
@ptabor - does this need to be ported to https://github.com/etcd-io/etcd/tree/release-3.5?

**Comment by ptabor** [comment]
> @ptabor - does this need to be ported to https://github.com/etcd-io/etcd/tree/release-3.5?

Changing protos (so public wire formats) is rather unexpected during a minor version lifetime, so I don't think there are strong reasons to backport (but also costs are low).

**Comment by nate-double-u** [comment]
The last force push updates the output as per @ptabor's suggestion.

```
protodoc is finished.

The API References have NOT been automatically published to the website.

To publish the API References, copy the following files
  - Documentation/dev-guide/api_reference_v3.md
  - Documentation/dev-guide/api_concurrency_reference_v3.md
to the etcd-io/website repo under the /content/en/docs/next/dev-guide/ folder.
(https://github.com/etcd-io/website/tree/main/content/en/docs/next/dev-guide)

./genproto SUCCESS
```

**Comment by nate-double-u** [comment]
The last force-push is a copy edit:

```
protodoc is finished.

The API references have NOT been automatically published on the website.

To publish the API references, copy the following files
  - Documentation/dev-guide/api_reference_v3.md
  - Documentation/dev-guide/api_concurrency_reference_v3.md
to the etcd-io/website repo under the /content/en/docs/next/dev-guide/ folder.
(https://github.com/etcd-io/website/tree/main/content/en/docs/next/dev-guide)

./genproto SUCCESS
```

**Comment by nate-double-u** [comment]
Thanks for the feedback @ptabor & @chalin!

**Pull request merged**